libexec/bats-core/bats-exec-file

```diff
@@ -134,64 +134,114 @@ function teardown_file() {
   return 0
 }
 
+bats_parallel_exec_test() { # $1 = test_output_dir, the rest is forwarded to bats-exec-test
+  test_output_dir="$1"
+  mkdir -p "$test_output_dir"
+  shift
+
+  # store the pid for waiting on finish
+  printf "%d" "$$" > "$test_output_dir/pid"
+
+  bats-exec-test "$@" 2>"$test_output_dir/stderr" >"$test_output_dir/stdout"
+
+  # store the result for marking end and retrieving errors
+  printf "%d" "$?" > "$test_output_dir/exitcode"
+}
+
+# make visible for GNU parallel
+export -f bats_parallel_exec_test
+
+bats_run_tests_in_parallel() {
+  for test_name in "${tests_to_run[@]}"; do
+    # Only handle non-empty lines
+    if [[ $test_name ]]; then
+      ((++test_number_in_suite))
+      ((++test_number_in_file))
+      # parallelize across files using parallel's semaphore mode
+      # shellcheck disable=SC2154 #bats_parallel_args is inherited from exec-suite
+      parallel --semaphore -j "$num_jobs" "${bats_parallel_args[@]}" -- bats_parallel_exec_test "$output_folder/$test_number_in_suite" "${flags[@]}" "$filename" "$test_name" "$test_number_in_suite" "$test_number_in_file"
+    fi
+  done
+}
+
+# forward the stdin/stderr output of the test with the given number
+# this will block until the test was started and finished before printig anything
+# this will also return the test's exit code
+bats_forward_output_of_test() {
+  test_number="$1"
+  test_output_dir="$output_folder/$test_number"
+  # wait for test to start
+  until [ -e "$test_output_dir/pid" ]
+  do
+    sleep 1
+  done
+  # wait for the test to be finished if it hasn't yet
+  if [[ ! -e "$test_output_dir/exitcode" ]]; then
+    # get the test's pid
+    test_pid=$(cat "$test_output_dir/pid")
+    # GNU parallel semaphore's subprocess will not be our child, 
+    # we cannot use `wait` but we can `tail --pid`!
+    if ! tail --pid="$test_pid" -f /dev/null 2>/dev/null; then
+      # when tail exits with non null, it does not support --pid 
+      # so we have to fall back to polling
+      until [ -e "$test_output_dir/exitcode" ]
+      do
+        sleep 1
```

Comment:
This looks like it could hang forever if there's a bug somewhere in bats. Should we put a timeout in here? Or logging somehow? Not sure either are good ideas...

Reply:
That's a tough one.

Firstly, there are two places where we could block: Waiting for the test to start and waiting for it to stop.

The starting is more under our control, so we could get away with timeouts there. However, the stopping is really hard because we can't decide on a timeout that will work for everyone out there. We might want to make this configurable then.
If we have the pid, we can also try to detect if the test process itself is still running and handle that differently.